test/solc/CommandLineInterface.cpp

```diff
@@ -278,6 +278,113 @@ BOOST_AUTO_TEST_CASE(cli_not_a_file)
 	);
 }
 
+BOOST_AUTO_TEST_CASE(linker)
+{
+	frontend::FileReader fileReader({},{},{});
+	fileReader.setSourceUnits({
+		{"bytecodeInput.bin", "565b6040516100579190610158565b60405180910390f35b600073__$b3157e226f2c4dddae135e6cab4ed4e747$__63b3de648b8360405"}
+	});
+
+	string hashString = "1234567890123456789012345678901234567890";
+	map<string, util::FixedHash<20>> libraries = {
+		{"test/libsolidity/semanticTests/libraries/stub.sol:L", FixedHash<20>("0x" + hashString)}
+	};
+
+	bool hasError = false;
+
+	CommandLineInterface::link(fileReader, libraries, hasError, cerr);
+	BOOST_TEST(!hasError);
+	BOOST_TEST(fileReader.sourceUnits().find("bytecodeInput.bin")->second.find(hashString) != string::npos);
+}
+
+BOOST_AUTO_TEST_CASE(linker_unmatched_markers_no_end)
+{
+	frontend::FileReader fileReader({},{},{});
+	fileReader.setSourceUnits({
+		{"bytecodeInput.bin", "565b6040516100579190610158565b60405180910390f35b600073__$b3157e226f2c4dddae135e6cab4ed4e74763b3de648b8360405"}
```

Comment:
Can you add a few more test cases? I think that the code might not work correctly with some of them:

- Old-style placeholder:
    ```
    565b604051610057919__lib.sol:L_____________________________48b8360405
    ```
- Two starts, single end:
    ```
    565b6040516100579190__$b36__$b3157e226f2c4dddae135e6cab4ed4e747$__63b3de648b8360405
    ```
- Placeholder in a placeholder:
    ```
    565b6040516100579190__$b36__$b3157e226f2c4dddae135e6cab4ed4e747$__63b3de$__648b8360405
    ```
- Double-sided marker:
    ```
    565b604051610057919__$__157e226f2c4dddae135e6cab4ed4e7__$__48b8360405
    ```
    I think this was valid in the old implementation and would be treated as an old-style placeholder.
- Marker without `$`:
    ```
    565b604051610057919___b3157e226f2c4dddae135e6cab4ed4e747___48b8360405
    ```
    Also technically a valid old-style placeholder.

Reply:
Should we be supporting the placeholder in a placeholder or double-sided marker cases?

Reply:
The double-sized one could be a valid old-style placeholder so yes, unless we drop support for old-style placeholders. The `__` on the inside basically counts as part of the library identifier in that case.
But I'm mentioning that case mostly because it's a corner case that looks like it could easily confuse regex based detection.

As for nested ones, if the outer `__` pair has the correct distance between markers, then again, it could be a valid old-style placeholder. But if it's the inner one that has the right length then the outer one should be treated as unmatched markers and result in an error.